src/Data/Binary/Serialise/CBOR/Tutorial.hs

```diff
@@ -16,33 +16,159 @@ module Data.Binary.Serialise.CBOR.Tutorial
+constructor.
+
+*** A note on interoperability
+
+@binary-serialise-cbor@ is intended primarily as a /serialisation/ library for
+Haskell values. That is, a means of stably storing Haskell values for later
+reading by @binary-serialise-cbor@. While it uses the CBOR encoding format, the
+library doesn't /not/ primarily aim to facilitate serialisation and
+deserialisation across different CBOR implementations.
+
+If you want to use @binary-serialise-cbor@ to serialise\/deserialise values
+for\/from another CBOR implementation (either in Haskell or another language),
+you should keep a few things in mind,
+
+1. The 'Serialise' instances for some "basic" Haskell types (e.g. 'Maybe',
+   'Data.ByteString.ByteString') don't carry a tag, in contrast to common
+   convention. This is an intentional design decision to minimize encoding size.
+
+2. The library reserves the right to change encodings in
+   non-backwards-compatible ways across super-major versions.
+
+3. While the library tries to use standard encodings in its instances wherever possible,
+   these instances aren't guaranteed to implement all valid variants of the
+   encodings in the specification. For instance, the 'UTCTime' instance only
+   implements a small subset of the encodings described by the Extended Date
+   RFC.
 
 -}
 
+
 {- $encoding
 
-Lorem ipsum...
+During encoding, abstract CBOR resentations are represented by the
+'Data.Binary.Serialise.CBOR.Encoding.Tokens' type. Such a representation can be
+efficiently built using the 'Data.Binary.Serialise.CBOR.Encoding.Encoding'
+'Monoid'.
+
+For instance, to implement an encoder for the @Animal@ type above we might write,
+
+> encodeAnimal :: Animal -> Encoding
+> encodeAnimal (HoppingAnimal name height) =
+>     encodeListLen 3 <> encodeTag 0 <> encodeString name <> encodeInt height
+> encodeAnimal (WalkingAnimal name speed) =
+>     encodeListLen 3 <> encodeTag 1 <> encodeString name <> encodeInt speed
+
+Here we see that each encoding begins with a /length/, describing how many
+values belonging to our @Animal@ will follow. We then encode a /tag/, which
+identifies which constructor. We then encode the fields.
 
 -}
 
 {- $decoding
 
-Lorem ipsum...
+Decoding CBOR representations to Haskell values is done in the 'Decoder'
+'Monad'. We can write a 'Decoder' for the @Animal@ type defined above as
+follows,
+
+> decodeAnimal :: Decoder s Animal
+> decodeAnimal = do
+>     len <- decodeListLen
+>     tag <- decodeTag
+>     case (len, tag) of
+>       (3, 0) -> HoppingAnimal <$> decodeString <*> decodeInt
+>       (3, 1) -> WalkingAnimal <$> decodeString <*> decodeInt
+>       _      -> fail "invalid Animal encoding"
+
+Note that we scrutinize both the length and the tag to determine which encoding
+we are looking at. This allows us to, for instance, add fields without changing
+using a new tag by simply appending new fields to the end of the encoding and
+changing the length appropriately.
```

Comment:
I'm not so sure this is a great idea. It does not help with forward compat. The only way to help there afaics is to add support for ignoring trailing fields we don't expect. With this scheme using both tag and length, there's still no way to add new fields to the end of a record in a way that older versions of the code will cope with (by ignoring).

The only advantage to dispatching on tag + length is not having to add new tag numbers so often.

Or if we think this is a sensible scheme then we should discuss it fully.